### PR TITLE
Core/intrusive ptr for variable list

### PR DESCRIPTION
**Pull request by RiccardoRossi**
title tells it all ... i think this is the proper way to handle this.

however we need to be extra careful since this is a big change

### CODE REVIEW

**Comment by RiccardoRossi** [comment]
actually one option would be to simplify the modelpart constructors.

not a big deal since anyway only the model can call them... but we can take that decision

**Review by pooyan-dadvand — COMMENTED** [review]
find_global_nodal_elemental_neighbours.h is part of this PR?

**Comment by philbucher** [comment]
it does not compile

File: kratos/containers/variables_list.h [hunk 1]
```diff
@@ -150,7 +154,11 @@ namespace Kratos
 				std::equal(mVariables.begin(), mVariables.end(), r.mVariables.begin());
 		}
 
-
+		//public API of intrusive_ptr
+		unsigned int use_count() const noexcept
```
Comment (rubenzorrilla) [inline kratos/containers/variables_list.h:158]: Just for me to know, why do we want the `use_count()` to not throw exceptions?

Comment (pooyan-dadvand) [reply kratos/containers/variables_list.h:158]: It affects the performance and we know that we are not sending exception there.

Comment (RiccardoRossi) [reply kratos/containers/variables_list.h:158]: also it comes in the recipe ;-)

**Comment by RiccardoRossi** [comment]
> find_global_nodal_elemental_neighbours.h is part of this PR?

removed

**Comment by philbucher** [comment]
I tried this with a restart case and it was working, hence no objections from my side

**Comment by philbucher** [comment]
briefly looking at the changes I think they are ok

**Review by pooyan-dadvand — APPROVED** [review]
Hopefully this PR will fix a lot of issues in destructing modelparts!

**Comment by philbucher** [comment]
this is Riccardo

**Pull request merged**